src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -193,48 +200,88 @@ public ProgrammingExerciseService(ProgrammingExerciseRepository programmingExerc
-
-        connectAuxiliaryRepositoriesToExercise(programmingExercise);
 
-        programmingExerciseRepositoryService.setupExerciseTemplate(programmingExercise, exerciseCreator);
-        programmingSubmissionService.createInitialSubmissions(programmingExercise);
+        programmingExercise.generateAndSetProjectKey();
+        programmingExercise.setBranch(versionControl.getDefaultBranchOfArtemis());
+        ProgrammingExercise createdExercise;
+
+        try {
+            // We execute the following steps in a transaction to be able to roll back in the case of errors.
+            // As a transactional method cannot be used in the same class, we have to use this workaround.
+            createdExercise = transactionTemplate.execute(new TransactionCallback<ProgrammingExercise>() {
+
+                @Override
+                public ProgrammingExercise doInTransaction(@NotNull TransactionStatus status) {
+                    try {
+                        programmingExerciseRepositoryService.createRepositoriesForNewExercise(programmingExercise);
+                    }
+                    catch (GitAPIException e) {
+                        log.error("An exception occurred while creating repositories for new exercise.", e);
+                        throw new RuntimeException(e);
+                    }
+                    initParticipations(programmingExercise);
+                    setURLsAndBuildPlanIDsForNewExercise(programmingExercise);
+
+                    // Save participations to get the ids required for the webhooks
+                    connectBaseParticipationsToExerciseAndSave(programmingExercise);
+
+                    connectAuxiliaryRepositoriesToExercise(programmingExercise);
+
+                    try {
+                        programmingExerciseRepositoryService.setupExerciseTemplate(programmingExercise, exerciseCreator);
+                    }
+                    catch (GitAPIException e) {
+                        log.error("An exception occurred while setting up the exercise template for new programming exercise.", e);
+                        throw new RuntimeException(e);
+                    }
+                    programmingSubmissionService.createInitialSubmissions(programmingExercise);
+
+                    // Save programming exercise to prevent transient exception
+                    ProgrammingExercise savedProgrammingExercise = programmingExerciseRepository.save(programmingExercise);
+
+                    channelService.createExerciseChannel(savedProgrammingExercise, Optional.ofNullable(programmingExercise.getChannelName()));
+
+                    setupBuildPlansForNewExercise(savedProgrammingExercise);
+                    return programmingExerciseRepository.saveAndFlush(savedProgrammingExercise);
+                }
+            });
+        }
+        catch (RuntimeException e) {
+            throw (GitAPIException) e.getCause();
+        }
```

Comment:
I’m not sure if this is safe. Other types of runtime exception might be thrown somewhere. What do you think about creating a
```java
private static final class WrappedGitApiException extends RuntimeException {
    WrappedGitApiException(GitApiException cause) {
        super(cause);
    }
}
```
and only catching that type here?
```suggestion
        catch (WrappedGitApiException e) {
            throw (GitAPIException) e.getCause();
        }
```

Reply:
Yes, I already had in mind, that there might be a problem, if other exceptions are thrown.

Your idea sounds good 👍 
I will implement it later. Thanks!